### PR TITLE
UI audio logic unit tests

### LINKED ISSUE #494: [FEATURE] UI Audio Logic Unit Tests (GUT):  Navigation Triggering
## Sub-Issue 3: GUT Unit Testing Suite
**Title**: `[FEATURE] UI Audio Logic Unit Tests (GUT)`
**Labels**: `testing`

**Description**:
Create a comprehensive test suite using the **GUT** plugin to verify the `AudioManager` correctly interprets input events as audio triggers.

### Test Plan 1: Navigation Triggering
* **File**: `res://test/unit/test_audio_navigation.gd`
* **Logic**:
    * Initialize `AudioManager`.
    * Simulate a `ui_up` InputEventKey[cite: 75, 120].
    * **Assertion**: Verify that the internal navigation `AudioStreamPlayer` is `is_playing()` and its `stream.resource_path` matches `ui_navigation.wav`.

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

# Pull Request Summary: Architectural Testing & Reliability Fixes for Audio UI

This PR introduces an automated testing suite for `AudioSettings` and implements significant resilience improvements for the `AudioManager` and `VolumeSlider` components. The primary objective is to transition from brittle, manually checked UI tests to a robust, API-driven test suite suitable for CI environments.

---

### 1. Architectural Testing Suite (`test_audio_settings_interaction.gd`)
* **GUT Integration:** Established a comprehensive test suite using the GUT framework.
* **Focus-Gate Validation:** Added stress tests to verify that audio triggers only fire when the UI component has focus, validating the logic in `audio_settings.gd`.
* **Resilience Testing:** Introduced boundary and stress tests to ensure:
    * Rapid UI interaction (spamming) does not overwhelm the audio pool.
    * Out-of-range volume values and invalid SFX keys are handled gracefully without crashing.
* **State Management:** Mocked `Globals.previous_scene` during test execution to prevent unintended scene changes (restarts) during UI interaction tests.

### 2. AudioManager Refactoring (`audio_manager.gd`)
* **Diagnostic APIs:** Added `is_sfx_playing()` and `get_active_sfx_stream_path()` to expose internal state to tests.
    * *Note:* These are explicitly commented as `## [DIAGNOSTIC]` to signal their appropriate use.
* **Determinism:** Updated `get_active_sfx_stream_path()` to return the most recently played SFX, ensuring tests have a deterministic target rather than an arbitrary pool index.

### 3. Component Resilience (`volume_slider.gd`)
* **Signal Integrity:** Updated the volume logic to allow programmatic updates via `_on_value_changed()` (explicitly called by tests), bypassing issues where `slider.value` property setters do not emit signals.
* **Input Clamping:** Implemented `clamp()` logic in `_on_value_changed` to ensure slider values always stay within valid bounds, improving resilience against invalid user or test inputs.

### 4. Repository Health
* **Metadata Cleanup:** Updated `.gitignore` to ignore engine-generated `*.uid` files, eliminating "noisy" diffs and potential merge conflicts.

---

## 🛠 Technical Fixes

* **Race Conditions:** Resolved crashes occurring when calling methods on nodes currently being freed by using `is_instance_valid(audio_instance)` checks.
* **Brittle Assertions:** Replaced hardcoded string substrings (e.g., `"check"`) with named constants (`SFX_CHECK_KEYWORD`) to ensure tests remain valid even if asset file names change.
* **Async Stability:** Added `await get_tree().create_timer(0.5).timeout` buffers in stress tests to ensure the audio system has time to reconcile playback state before assertions run.

---

## 🧪 Testing Status

| Category | Status | Notes |
| :--- | :--- | :--- |
| **Interaction Tests** | ✅ Passed | Mute toggles, reset functionality, and back navigation. |
| **Regression/Spam** | ✅ Passed | Verified pool usage under heavy load. |
| **Boundary/Resilience** | ✅ Passed | Invalid inputs, extreme slider values, corrupted states. |
| **Focus Gate** | ✅ Passed | Confirmed silent-when-unfocused logic. |

**Final Verification:** 12/12 Tests Passed.

---

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [ ] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [ ] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [ ] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [ ] Any new unit tests added? (Link to test scene if yes)
- [ ] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [ ] Code follows Godot style guide (e.g., snake_case for variables)
- [ ] No console errors in editor/output
- [ ] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Add public AudioManager helpers for inspecting active SFX and introduce a comprehensive UI interaction test suite for audio settings behavior.

New Features:
- Expose AudioManager APIs to query whether any SFX is playing and to retrieve the active SFX stream path.

Enhancements:
- Add architectural GUT tests covering audio settings UI interactions, including mute toggles, reset behavior, slider updates, and stress/resilience scenarios.

CI:
- Update Codecov GitHub Action to a newer pinned commit in the browser test workflow.

Tests:
- Add regression-focused tests to validate audio pool behavior, focus-gated SFX playback, slider boundary handling, invalid input resilience, and reset recovery for audio settings UI.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed volume slider to properly clamp input values within valid range, preventing out-of-bounds audio settings.

* **Tests**
  * Added comprehensive test suite for audio settings interactions, validating mute toggling, volume slider behavior, and reset functionality.

* **Chores**
  * Updated workflow action dependency.

---

**Bots/AI Contributions Summary for PR #746**

This PR adds public diagnostic helpers (`is_sfx_playing()`, `get_active_sfx_stream_path()`) to `AudioManager`, implements comprehensive GUT tests for UI audio interactions (mute toggles, sliders, reset behavior, focus-gating, boundary/stress cases), includes a minor bug fix for volume slider clamping, and updates the Codecov GitHub Action. It received strong support from automated bots and AI tools for dependency management, PR summarization, review, and code quality analysis.

### Automated Bots & AI Tools

- **@dependabot[bot]**: Handled the dependency update by bumping `codecov/codecov-action` from 6.0.1 to 7.0.0 (major version) in the browser test workflow, including the automated commit and related merge. This improves coverage reporting reliability and security.

- **@sourcery-ai**: Provided the primary structured PR summary and reviewer's guide. Highlighted new AudioManager APIs, the extensive GUT test suite (UI interactions, resilience, stress scenarios), CI updates, and linked the work to issue #494. Also contributed to title and description refinement.

- **@coderabbitai**: Delivered a focused summary covering bug fixes (volume slider clamping), new comprehensive test suite for audio settings interactions, and maintenance chores (workflow dependency update).

- **@deepsource-io**: Performed automated static code analysis and code review on the changes (AudioManager extensions, test suite, CI workflow). Provided an overall grade across Security, Reliability, Complexity, and Hygiene categories, along with inline comments and a full review report.

These tools enhanced test coverage documentation, reviewer guidance, dependency security, and overall code health validation.

### Human Maintainers

- **@ikostan**: Primary contributor and PR author. Led the full implementation, including new diagnostic APIs on `AudioManager`, volume slider input clamping fix, comprehensive GUT test suite (`test_audio_settings_interaction.gd`) covering interaction scenarios, focus-gating, resilience, and stress tests, test harness helpers, sequence diagrams, milestone integration, and CI workflow updates.

---

### CODE REVIEW

**Comment by ikostan** [comment]
@sourcery-ai review

**Comment by ikostan** [comment]
> * The tests reach into AudioManager internals (e.g., `_sfx_pool`, `get_active_sfx_playback_count`) and controller `_on_*` handlers directly, which makes them tightly coupled to implementation details—consider driving behavior via the public UI/API or signals where possible to keep the tests resilient to refactors.

Already addressed.

**Comment by ikostan** [comment]
> * `test_back_button_triggers_exit` currently calls `queue_free()` on `audio_instance` directly instead of exercising the actual back-button handler, so it isn't validating the real interaction path; consider invoking the same method or signal the UI uses for back navigation.

I removed the test.

**Comment by ikostan** [comment]
> * Check whether `test/gut/test_audio_settings_interaction.gd.uid` needs to be versioned, as committing engine-generated UID/metadata files can cause noisy diffs and merge conflicts if they are regenerated frequently.

Addressed

**Comment by ikostan** [comment]
@sourcery-ai review

File: test/gut/test_audio_settings_interaction.gd [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+# 3. REGRESSION & RESILIENCE
+# ==========================================================================
+
+## Validates extreme volume values are handled correctly (boundary testing).
+## :rtype: void
+func test_slider_extreme_values() -> void:
+	_clear_pool_players()
+	var slider: HSlider = audio_instance.master_slider
+	
+	slider.value = 0.0
```
[the remark this answers is not in the document]
Comment (ikostan) [reply test/gut/test_audio_settings_interaction.gd:179]: Can you test this issue again and make sure it was fully addressed?

File: test/gut/test_audio_settings_interaction.gd [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+
+## Validates audio pool does not leak under heavy interaction stress.
+## :rtype: void
+func test_audio_pool_no_leak_under_stress() -> void:
+	_clear_pool_players()
+	var initial_count := AudioManager.get_active_sfx_playback_count()
+	
+	for i in range(20):
+		audio_instance._on_music_mute_toggled(i % 2 == 0)
```
[the remark this answers is not in the document]
Comment (ikostan) [reply test/gut/test_audio_settings_interaction.gd:230]: Could you please verify this issue again for me?

File: scripts/managers/audio_manager.gd [hunk 3]
```diff
@@ -477,3 +477,21 @@ func stop_all_sfx() -> void:
...
+		if player.playing:
+			return true
+	return false
+
+
+## Public API: Returns the resource path of the currently playing SFX.
+## @return String: Resource path of the active stream, or empty if silent.
+func get_active_sfx_stream_path() -> String:
+	for player in _sfx_pool:
+		if player.playing and player.stream:
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/managers/audio_manager.gd]: The reviewer is correct: returning an "arbitrary" sound is bad practice because it makes your tests flaky (they might pass or fail depending on the order in which sounds were loaded into the pool).

To resolve this, you should make the behavior deterministic. Instead of just returning the "first one it finds," update the logic to return the most recently played SFX by iterating through the pool in reverse order. This turns an "arbitrary" result into an "explicit rule."

File: scripts/managers/audio_manager.gd [hunk 4]
```diff
@@ -477,3 +477,21 @@ func stop_all_sfx() -> void:
...
 		player.stream = null
+
+
+## Checks if any SFX is currently playing.
+## @return bool
+func is_sfx_playing() -> bool:
+	for player in _sfx_pool:
+		if player.playing:
+			return true
+	return false
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/managers/audio_manager.gd]: Can you check it again?

**Comment by ikostan** [comment]
@sourcery-ai guide
